jena-core/src/main/java/org/apache/jena/reasoner/rulesys/impl/SafeTripleIterator.java

```diff
@@ -0,0 +1,56 @@
+package org.apache.jena.reasoner.rulesys.impl;
+
+import org.apache.jena.graph.Triple;
+import org.apache.jena.reasoner.TriplePattern;
+import org.apache.jena.reasoner.rulesys.BackwardRuleInfGraphI;
+import org.apache.jena.util.iterator.ClosableIterator;
+import org.apache.jena.util.iterator.ExtendedIterator;
+
+/**
+ * Wrapper around triple searches using in the back chainer to attempt a restart
+ * in the case of a failure such as cross-transaction query.
+ */
+public class SafeTripleIterator implements ClosableIterator<Triple> {
+
+    protected ExtendedIterator<Triple> matchIterator;
+    protected long offset = 0;
+    protected TriplePattern goal;
+    protected BackwardRuleInfGraphI infGraph;
+
+    public SafeTripleIterator(LPInterpreter interpreter, TriplePattern goal) {
+        this.infGraph = interpreter.getEngine().getInfGraph();
+        this.goal = goal;
+        restart();
+    }
+
+    protected void restart() {
+        matchIterator = infGraph.findDataMatches(goal);
+        for (int i = 0; i < offset; i++) matchIterator.next();
```

Comment:
Doesn't this needs to defend against the case where the data changes rendering the offset invalid i.e. it might no longer be possible to call `next()` the expected number of times resulting in a `NoSuchElementException`

Reply:
For the case of `restart()` under `next()` this is fine but for `restart()` under `hasNext()` you'd expect to get a `false` return not an exception

Reply:
If the data changes then there should have been either an explicit `rebind()` or an implicit `prepare()` call which should clear the tabled goals anyway. If you make changes to the data bypassing both of those then iterators will blow up anyway, don't think the restarting makes this any worse.